lldb/test/Shell/helper/stepper.py

```diff
@@ -0,0 +1,221 @@
+    for o in opts:
+        m = re.match('(\w+)=(.+)', o)
+        if not m:
+            print('Unrecognized option:', o)
+            continue
+        option, val = m.groups()
+        if option not in globals():
+            print('Unrecognized option:', option)
+            continue
+        print('Setting', option, '=', val)
+        globals()[option] = eval(val)
+
+
+def doit(dbg, cmd):
+    "Run a driver command."
+    print('::', cmd)
+    dbg.HandleCommand(cmd)
+    return False
+
+
+def doquit(dbg):
+    "Quit the stepper script interpreter."
+    exit(0)
+
+
+def should_stop_stepping(process):
+    "Decide whether we should stop stepping."
+    state = process.GetState()
+    if state in (lldb.eStateExited, lldb.eStateDetached):
+        print('Process has exited or has been detached, exiting...')
+        return True
+    if state in (lldb.eStateCrashed, lldb.eStateInvalid):
+        print('Process has crashed or is in an invalid state, exiting...')
+        return True
+    return False
+
+
+def alter_PC(dbg, process, cmd):
+    "Run a driver command that changes the PC. Return whether to stop stepping."
+    # Check the process state /after/ we step. Any time we advance the PC,
+    # the process state may change.
+    doit(dbg, cmd)
+    return should_stop_stepping(process)
+
+
+def return_from_frame(thread, frame):
+    print(':: Popping current frame...')
+    old_name = frame.GetFunctionName()
+    thread.StepOutOfFrame(frame)
+    new_frame = thread.GetSelectedFrame()
+    new_name = new_frame.GetFunctionName()
+    print(':: Transitioned from {} -> {}.'.format(old_name, new_name))
+    return True
+
+
+def __lldb_init_module(dbg, internal_dict):
+    parse_options()
+
+    # Make each debugger command synchronous.
+    dbg.SetAsync(False)
+
+    # Run the program and stop it when it reaches main().
+    if RUN_THE_PROCESS:
+        doit(dbg, 'breakpoint set -n main')
+        doit(dbg, 'run')
+
+    # Step through the program until it exits.
+    gen = 0
+    inspections = 0
+    target = dbg.GetSelectedTarget()
+    process = target.GetProcess()
+    visited_pc_counts = defaultdict(int)
+    while True:
+        gen += 1
+        print(':: Generation {} (# inspections = {})'.format(gen, inspections))
+
+        thread = process.GetSelectedThread()
+        frame = thread.GetSelectedFrame()
+
+        do_inspection = True
+
+        # Sometimes, lldb gets lost after stepping. This is rdar://70546777.
+        # Try to remind lldb where it is by running 'frame select'.
+        if not frame.GetFunctionName():
+            doit(dbg, 'frame select')
+
+        # Skip frames without valid line entries.
+        line_entry = frame.GetLineEntry()
+        if do_inspection and not line_entry.IsValid():
+            do_inspection = False
+
+        # Skip optimized frames if asked to do so.
+        if do_inspection and ONLY_VISIT_UNOPTIMIZED and \
+                str(frame).endswith(' [opt]'):
```

Comment:
TIL `SBFrame` has `IsArtificial` and `IsInlined` but not `IsOptimized`. Should we add this?

Reply:
Probably, yeah. I'll put a patch together.

Reply:
Ah, maybe SBFunction::GetIsOptimized is sufficient? I can see the logic behind that. Whether or not a function is inlined or tail-called depends on the frame. But if a function is optimized, that's true independent of which frame you pick.